kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflow.kt

```diff
@@ -18,107 +18,55 @@ package com.squareup.workflow
  * Typically represents a "view" of this workflow's input, current state, and children's renderings.
  * A workflow that represents a UI component may use a view model as its rendering type.
+ *
+ * @see StatefulWorkflow
+ * @see StatelessWorkflow
  */
-interface Workflow<in InputT : Any, StateT : Any, out OutputT : Any, out RenderingT : Any> {
-
-  /**
-   * Called when the state machine is first started to get the initial state.
-   */
-  fun initialState(input: InputT): StateT
-
-  /**
-   * Called whenever [WorkflowContext.compose] is about to get a new [input][InputT] value, to allow
-   * the workflow to modify its state in response. This method is called eagerly: `old` and `new` might
-   * be the same value, so it is up to implementing code to perform any diffing if desired.
-   *
-   * Default implementation does nothing.
-   */
-  fun onInputChanged(
-    old: InputT,
-    new: InputT,
-    state: StateT
-  ): StateT = state
-
-  /**
-   * Called at least once† any time one of the following things happens:
-   *  - This workflow's [input] changes (via the parent passing a different one in).
-   *  - This workflow's [state] changes.
-   *  - A child workflow's state changes.
-   *
-   * **Never call this method directly.** To get the rendering from a child workflow, pass the child
-   * and any required input to [WorkflowContext.compose].
-   *
-   * This method *should not* have any side effects, and in particular should not do anything that
-   * blocks the current thread. It may be called multiple times for the same state. It must do all its
-   * work by calling methods on [context].
-   *
-   * _† This method is guaranteed to be called *at least* once for every state, but may be called
-   * multiple times. Allowing this method to be invoked multiple times makes the internals simpler._
-   */
-  fun compose(
-    input: InputT,
-    state: StateT,
-    context: WorkflowContext<StateT, OutputT>
-  ): RenderingT
-
-  /**
-   * Called whenever the state changes to generate a new [Snapshot] of the state.
-   *
-   * **Snapshots must be lazy.**
-   *
-   * Serialization must not be done at the time this method is called,
-   * since the state will be snapshotted frequently but the serialized form may only be needed very
-   * rarely.
-   *
-   * If the workflow does not have any state, or should always be started from scratch, return
-   * [Snapshot.EMPTY] from this method.
-   *
-   * @see restoreState
-   */
-  fun snapshotState(state: StateT): Snapshot
+interface Workflow<in InputT : Any, out OutputT : Any, out RenderingT : Any> {
 
-  /**
-   * Deserialize a state value from a [Snapshot] previously created with [snapshotState].
-   *
-   * If the workflow should always be started from scratch, this method can just ignore the snapshot
-   * and return the initial state.
-   *
-   * @see snapshotState
-   */
-  fun restoreState(snapshot: Snapshot): StateT
+  fun asStatefulWorkflow(): StatefulWorkflow<InputT, *, OutputT, RenderingT>
```

Comment:
Provides a `StatefulWorkflow` view of the receiver. Necessary because `StatefulWorkflow` is the common API required for `WorkflowContext.compose()` to do its work.

Reply:
And this is why i always leave `TODO kdoc` comments everywhere…